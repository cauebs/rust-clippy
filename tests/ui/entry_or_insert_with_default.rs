#![warn(clippy::entry_or_insert_with_default)]

use std::collections::{BTreeMap, HashMap};

fn btree(map: &mut BTreeMap<i32, i32>) {
    map.entry(42).or_insert_with(Default::default);
    map.entry(42).or_insert_with(i32::default);
    map.entry(42).or_default();
}

fn hash(map: &mut HashMap<i32, i32>) {
    map.entry(42).or_insert_with(Default::default);
    map.entry(42).or_insert_with(i32::default);
    map.entry(42).or_default();
}

struct InformalDefault;

impl InformalDefault {
    fn default() -> Self {
        InformalDefault
    }
}

fn not_default_btree(map: &mut BTreeMap<i32, InformalDefault>) {
    map.entry(42).or_insert_with(InformalDefault::default);
}

fn not_default_hash(map: &mut HashMap<i32, InformalDefault>) {
    map.entry(42).or_insert_with(InformalDefault::default);
}

#[derive(Default)]
struct NotAnEntry;

impl NotAnEntry {
    fn or_insert_with(self, f: impl Fn() -> Self) {}
}

fn main() {
    NotAnEntry.or_insert_with(Default::default);
    NotAnEntry.or_insert_with(NotAnEntry::default);
}
