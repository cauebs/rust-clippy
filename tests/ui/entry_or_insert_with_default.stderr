error: use of `or_insert_with` with `Default::default`
  --> $DIR/entry_or_insert_with_default.rs:6:19
   |
LL |     map.entry(42).or_insert_with(Default::default);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `or_default()`
   |
   = note: `-D clippy::entry-or-insert-with-default` implied by `-D warnings`

error: use of `or_insert_with` with `Default::default`
  --> $DIR/entry_or_insert_with_default.rs:7:19
   |
LL |     map.entry(42).or_insert_with(i32::default);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `or_default()`

error: use of `or_insert_with` with `Default::default`
  --> $DIR/entry_or_insert_with_default.rs:12:19
   |
LL |     map.entry(42).or_insert_with(Default::default);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `or_default()`

error: use of `or_insert_with` with `Default::default`
  --> $DIR/entry_or_insert_with_default.rs:13:19
   |
LL |     map.entry(42).or_insert_with(i32::default);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `or_default()`

error: aborting due to 4 previous errors

